src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Ssl.cs

```diff
@@ -325,6 +327,33 @@ internal enum SslErrorCode
     }
 }
 
+namespace System.Runtime.InteropServices.Marshalling
+{
+    [CustomMarshaller(typeof(SafeX509Handle), MarshalMode.ManagedToUnmanagedOut, typeof(NullableX509HandleMarshaller.Out))]
+    internal static class NullableX509HandleMarshaller
+    {
+        public struct Out
+        {
+            private SafeX509Handle? _handle;
+
+            public void FromUnmanaged(IntPtr value)
+            {
+                if (value != IntPtr.Zero)
+                {
+                    _handle = new SafeX509Handle();
+                    Marshal.InitHandle(_handle, value);
+                }
+            }
+
+            public SafeX509Handle? ToManaged() => _handle;
+
+            public void Free()
+            {
+            }
+        }
+    }
+}
```

Comment:
Is using a custom marshaller necessary here? It could just be:
```C#
[LibraryImport(Libraries.CryptoNative, EntryPoint = "CryptoNative_SslGetPeerCertificate")]
private static partial IntPtr SslGetPeerCertificate_Private(SafeSslHandle ssl);

internal static SafeX509Handle? SslGetPeerCertificate(SafeSslHandle ssl)
{
    IntPtr ptr = SslGetPeerCertificate_Private(ssl);
    if (ptr != IntPtr.Zero)
    {
        var handle = new SafeX509Handle();
        handle.InitHandle(ptr);
        return handle;
    }
    return null;
}
```
What advantages does the custom marshaling in this scenario provide?

Also, as was discussed, both this and the PR version suffer from one of the issues the default SafeHandle marshaling intends to address, that of the allocation of the SafeHandle itself failing.  Now if the handle allocation fails, we'll leak the pointer.  Presumably both the proposed marshaler and the above would need to deal with that, either in the above with a try/catch or in the marshaler by implementing Free?

Reply:
And, actually, looking at the call site, the SafeHandle ends up getting disposed of the in same method that consumes it.  We don't really need the SafeHandle at all in that case.  If we're concerned about the SafeHandle allocation on this code path, the P/Invoke can just return IntPtr and the consumer can just use it as an IntPtr, which is what it already wants in order to construct an X509Certificate2 from it.

Reply:
I switched to the IntPtr as it avoid the safe handle in both cases. I check and the native calls `SSL_get1_peer_certificate` that already grabs reference on the underlying `X509`. So we need to free it if allocation of the `X509Certificate` fails for what ever reason.